### ISSUE TITLE
[sdk/go] Allow reading the resource options that are set out of a `ResourceOption` 

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#11698]
Component authors are handed a `pulumi.ResourceOption` as part of implementing `Construct` or their own component resource.  

For example:

```go
type ConstructFunc func(ctx *pulumi.Context, typ, name string, inputs ConstructInputs,
	options pulumi.ResourceOption) (*ConstructResult, error)
```

However, `pulumi.ResourceOption` does not expose what resource options are actually set.

Internally, this data is available in the (unexported) `resourceOptions` struct that actually implements the `pulumi.ResourceOption` interface.  We will need to find a way to allow users to see thse value - either by exporting that struct, or by providing accessors on `pulumi.ResourceOption` to get the underlying values of every resource option.

This is necessary so that component authors can decide if and when to pass resource options down into children.  While many resource options are automatically inherited, some are not, and component authors need to be able to access them.  In other Pulumi languages, they can (and do).

**Comment by RobbieMcKinstry** [comment]
I would rather we add access on the interface, instead of exporting the struct. That way, we can swap out different implementors if needed.

**Comment by justinvp** [comment]
The "snapshot" API is along the lines of what I had in mind. I wonder if we should call it just `ResourceOptions`, w/out "Snapshot" in the name, and if it should look something like:

```go
func NewResourceOptions(opts ...ResourceOption) ResourceOptions {
}

type ResourceOptions struct {
    // ...
}

func (opts ResourceOptions) Options() []ResourceOption {
	// convert struct to slice
}
```

(And similar for `InvokeOptions`).

**Comment by justinvp** [comment]
Some hacky workaround ideas in the meantime:

1. For MLCs: Grab the resource option values that you need from the raw gRPC `ConstructRequest` and then pass those values through to your internal component factories so you can later extract them. Basically do your own equivalent of https://github.com/pulumi/pulumi/blob/6248c086aef9ce15e1ac69f53ac6ee05913b8d27/sdk/go/pulumi/provider.go#L88-L119 and pass the individual values through. There is some trickiness for `Providers` and `Parent` since `createProviderResource` and `newDependencyResource` aren't exported, but I think it'd be possible to recreate those just with publicly exported APIs.

2. I wouldn't recommend it, but another workaround would be to fork the Go SDK and provide accessors for the unexported struct of options, or something along the lines of the design discussed above. But at that point, might as well consider contributing a PR that changes it in this repo.

**Comment by abhinav** [comment]
This was released for resource options in 3.57, and for invoke options in 3.58.

- [pulumi.NewResourceOptions](https://pkg.go.dev/github.com/pulumi/pulumi/sdk/v3/go/pulumi#NewResourceOptions) takes a collection of `ResourceOption` values and returns a [ResourceOptions struct](https://pkg.go.dev/github.com/pulumi/pulumi/sdk/v3/go/pulumi#ResourceOptions)
- [pulumi.NewInvokeOptions](https://pkg.go.dev/github.com/pulumi/pulumi/sdk/v3/go/pulumi#NewInvokeOptions) similarly takes `InvokeOption` values and returns a [InvokeOptions struct](https://pkg.go.dev/github.com/pulumi/pulumi/sdk/v3/go/pulumi#InvokeOptions).

Closing as done.

**Comment by EronWright** [comment]
@justinvp I observe that the `Options()` method (as you suggested [here](https://github.com/pulumi/pulumi/issues/11698#issuecomment-1374638771)) wasn't implemented, was that intentional? It may be valuable for constructing child resources.
